contracts/utils/Arrays.sol

```diff
@@ -23,55 +23,151 @@ library Arrays {
      * when executing this as part of a transaction. If the array being sorted is too large, the sort operation may
      * consume more gas than is available in a block, leading to potential DoS.
      */
+    function sort(
+        bytes32[] memory array,
+        function(bytes32, bytes32) pure returns (bool) comp
+    ) internal pure returns (bytes32[] memory) {
+        uint256 begin;
+        uint256 end;
+        /// @solidity memory-safe-assembly
+        assembly {
+            begin := add(array, 0x20)
+            end := add(begin, mul(mload(array), 0x20))
+        }
+        _quickSort(begin, end, comp);
+        return array;
+    }
+
+    /**
+     * @dev Variant of {sort} that sorts an array of bytes32 in increassing order.
+     */
+    function sort(bytes32[] memory array) internal pure returns (bytes32[] memory) {
+        return sort(array, _defaultComp);
+    }
+
+    /**
+     * @dev Variant of {sort} that sorts an array of address following a provided comparator function.
+     */
+    function sort(
+        address[] memory array,
+        function(address, address) pure returns (bool) comp
+    ) internal pure returns (address[] memory) {
+        sort(_castToBytes32Array(array), _castToBytes32Comp(comp));
+        return array;
+    }
+
+    /**
+     * @dev Variant of {sort} that sorts an array of address in increassing order.
+     */
+    function sort(address[] memory array) internal pure returns (address[] memory) {
+        sort(_castToBytes32Array(array), _defaultComp);
+        return array;
+    }
+
+    /**
+     * @dev Variant of {sort} that sorts an array of uint256 following a provided comparator function.
+     */
+    function sort(
+        uint256[] memory array,
+        function(uint256, uint256) pure returns (bool) comp
+    ) internal pure returns (uint256[] memory) {
+        sort(_castToBytes32Array(array), _castToBytes32Comp(comp));
+        return array;
+    }
+
+    /**
+     * @dev Variant of {sort} that sorts an array of uint256 in increassing order.
+     */
     function sort(uint256[] memory array) internal pure returns (uint256[] memory) {
-        _quickSort(array, 0, array.length);
+        sort(_castToBytes32Array(array), _defaultComp);
         return array;
     }
 
     /**
-     * @dev Performs a quick sort on an array in memory. The array is sorted in increasing order.
+     * @dev Performs a quick sort of a segment of memory. The segment sorted starts at `begin` (inclusive), and stops
+     * at end (exclusive). Sorting follows the `comp` comparator.
      *
-     * Invariant: `i <= j <= array.length`. This is the case when initially called by {sort} and is preserved in
-     * subcalls.
+     * Invariant: `begin <= end`. This is the case when initially called by {sort} and is preserved in subcalls.
```

Comment:
```suggestion
     * Invariant: `begin <= end <= array.length`. This is the case when initially called by {sort} and is preserved in subcalls.
```

I think it's important to keep `array.length` because we're using `unsafeMemoryAccess`. If we don't mention it at all, it's dangerous for someone to provide an arbitrary end. Actually, I'm leaning more towards a note:

```suggestion
     * Invariant: `begin <= end`. This is the case when initially called by {sort} and is preserved in subcalls.
     *
     * IMPORTANT: Memory locations between `begin` and `end` are not validated/zeroed. This function should
     * be used only if the limits are within a memory array.
```

Reply:
- `end` is a memory pointer, it is not smaller than the length.

I'm ok with the second proposal is ok to me (though I would rewrite "validated/zeroed" as "allocated".

Reply:
Also, this is a private function, so we should consider that documentation is mostly (only?) for us. Until recently, we did not even document these private functions.

Reply:
Yeah this is true, although the original `sort` documentation is not as clear as this one. I think that's ok, but I'd rather be clear since all other functions are entry points to this